kong/pdk/client.lua

```diff
@@ -121,6 +127,71 @@ local function new(self)
   end
 
 
+  ---
+  -- Returns the credentials of the currently authenticated consumer.
+  -- If not set yet, it returns `nil`.
+  -- @function kong.client.get_credential
+  -- @phases access, header_filter, body_filter, log
+  -- @return the authenticated credential
+  -- @usage
+  -- local credential = kong.client.get_credential()
+  -- if credential then
+  --   consumer_id = credential.consumer_id
+  -- else
+  --   -- request not authenticated yet
+  -- end
+  function _CLIENT.get_credential()
+    check_phase(AUTH_AND_LATER)
+
+    return ngx.ctx.authenticated_credential
+  end
+
+
+  ---
+  -- Returns the `consumer` entity of the currently authenticated consumer.
+  -- If not set yet, it returns `nil`.
+  -- @function kong.client.get_consumer
+  -- @phases access, header_filter, body_filter, log
+  -- @treturn table the authenticated consumer entity
+  -- @usage
+  -- local consumer = kong.client.get_consumer()
+  -- if consumer then
+  --   consumer_id = consumer.id
+  -- else
+  --   -- request not authenticated yet, or a credential
+  --   -- without a consumer (external auth)
+  -- end
+  function _CLIENT.get_consumer()
+    check_phase(AUTH_AND_LATER)
+
+    return ngx.ctx.authenticated_consumer
+  end
+
+
+  ---
+  -- Sets the authenticated consumer (and credential) for the current request.
+  -- @function kong.client.set_consumer
+  -- @phases access
+  -- @tparam table consumer The consumer to set, this can be `nil`.
+  -- @tparam table credential The credential to set, this can be `nil`.
+  -- @usage
+  -- -- assuming `credential` and `consumer` have been set by some authentication code
+  -- kong.client.set_consumer(consumer, credentials)
+  function _CLIENT.set_consumer(consumer, credential)
```

Comment:
Where is `set_credential`? E.g. OpenID Connect Plugin can set `credential` without `consumer` for some reasons:
1. rate-limiting can then rate-limit by that
2. the `if conf.anonymous and ngx.ctx.authenticated_credential then ...` works with 3rd party authenticating plugins too.

Also I am not totally sure should we have `credential` here as a second argument, if we add `set_credential`.

Also I don't think these Kong concepts belong to `client`? Maybe `kong.consumer.set` and `kong.consumer.get` and `kong.credential.set` and `kong.credential.get`.

Reply:
I was wondering on `set_credential` as well. But updating the plugins (2nd commit) showed that this actually is ok. Because you should always set them both, if only a credential, then set the consumer to `nil`. Only makes the code more explicit.

wrt to `client` or `consumer` namespace, I think `client` is fine, auth is a downstream property imo.

Reply:
Yes, fair enough. As we approach the higher level functionality in PDK it is impossible to be very unopionionated. Still thinking where should we place these.`kong.auth` is another option. Should these functions also take care of setting headers?

Reply:
If consumer is sometimes optional should we then just have set_credential that takes consumer as second arg. Or do you see use case for consumer without credential? Maybe setting of anonymous consumer? But is it responsibility of these functions to set the upstream headers? Or should we patch core to do that?

Reply:
Setting the headers is a complex matter, because it involves removing, adding, and replacing headers. And the sets of headers differ per auth method.
That said, it is out of scope for this PR, I don't even think we can/should solve that now, since we'd first need to update the plugin-runloop to have something like an "auth" phase (basically moving the and/or logic for multiple-auth out of the individual plugins into core)

having independent set_consumer and set_credential is not a good idea imo.

Reply:
Yes, that's why I was suggesting having just `set_credential`, e.g. possibly `kong.auth.set_credential(credential or , consumer or nil)` Or perhaps some table to set this:*
```lua
{
  consumer = ...,
  credential = ...,
}
```

Because of all the comments here (e.g. regarding to runloop), it would probably make sense to stop and think about this for a moment. As with PDK we want proper backward and forward compatibility.